packages/react-dom/src/__tests__/ReactDOMFizzForm-test.js

```diff
@@ -360,4 +364,75 @@ describe('ReactDOMFizzForm', () => {
     expect(buttonRef.current.hasAttribute('formMethod')).toBe(false);
     expect(buttonRef.current.hasAttribute('formTarget')).toBe(false);
   });
+
+  // @gate enableFormActions
+  it('should replay a form action after hydration', async () => {
+    let foo;
+    function action(formData) {
+      foo = formData.get('foo');
+    }
+    function App() {
+      return (
+        <form action={action}>
+          <input type="text" name="foo" defaultValue="bar" />
+        </form>
+      );
+    }
+
+    const stream = await ReactDOMServer.renderToReadableStream(<App />);
+    await readIntoContainer(stream);
+
+    // Dispatch an event before hydration
+    submit(container.getElementsByTagName('form')[0]);
+
+    await act(async () => {
+      ReactDOMClient.hydrateRoot(container, <App />);
+    });
+
+    // It should've now been replayed
+    expect(foo).toBe('bar');
```

Comment:
make sure it's called just once?

make sure formData is captured at the time of submission?